src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs

```diff
@@ -434,6 +434,38 @@ public void VerifyCompilerAssemblySimpleNames()
                 allReferenceSimpleNames.Free();
             }
         }
+
+        [Fact]
+        public void AssemblyLoadingInNonDefaultContext_AnalyzerReferencesSystemCollectionsImmutable()
```

Comment:
I'm not certain this test covers the scenario I was thinking of. In the meeting notes I wrote:

> Imagine if they use Newtonsoft in their generator. ServiceHub host might have Newtonsoft loaded. What if the "compiler ALC" given to us by ServiceHub ends up giving us Newtonsoft.Json when we didn't want it?
> 
> We need a unit test for this. An assembly depends on some assembly "Util.dll" that the compiler DOES NOT USE but *would* be loaded by a "compiler ALC". The analyzer needs to reference a different, incompatible version of "util.dll" than the "compiler ALC" would provide. Then we verify that the DirectoryLoadContext loads the analyzer's version of the dependency rather than the version that the "compiler ALC" would give us.

I think this test doesn't cover the scenario described above, since System.Collections.Immutable is an explicit compiler dependency. I'll follow up when I manage to form a coherent thought on this subject.. it seems like the test you created has value, especially if it fails in the absence of the implementation change, but that we probably need more tests.

Reply:
The test I added is based on the error I actually saw when running Roslyn OOP on ServiceHub Core host: we are trying to load MS.CA.CSharp as an analyzer reference, the `DirectoryLoadContext` then delegated the load to `ALC.Default` by returning null in `DirectoryLoadContext.Load`, since MS.CA.CSharp is one of the compiler assemblies on our list.

I can definitely add the test you described in this PR as well, I didn't because I thought it's not directly related to the particular change in this PR.

Reply:
at this point I'm thinking it's OK to merge as long as Jared's feedback is addressed. I am thinking that I will follow up later on and add some more of the gnarly tests. We have another one which looks roughly like the following

- compile an analyzer against net13.
- Use valuetask in the analyzer's implementation.
- find out where valuetask got loaded from. if it's from the default context it's a failure.

Reply:
OK, added.